Update farm_map_google behavior contrib module for farmOS-map 2.x

symbioquine opened the pull request:
**Note: There ~~are two prerequisites~~ `is one more prerequisite` for this change to work correctly when merged;**

1. ~~https://github.com/mstenta/ol3-google-maps/commit/62c347b701a886470b594e0dc6049eac16dc4f88 needs to be pushed up to the NPM package [@farmos.org/farmos-olgm](https://www.npmjs.com/package/@farmos.org/farmos-olgm).~~
2. Some file - readme/license/etc - needs to be pushed to the new `2.x-release` branch manually in order to allow the checkout step in the Github action to work for the first release tag push.

**Why?** farmOS-map 2.x no longer contains core support for Google map
layers. Thus this contrib module will need to bundle that itself.

**What?** I'm converting this package to work similarly to
https://github.com/symbioquine/farmOS_vue_page_demo where
one branch (2.x) contains the unbuilt javascript and another
branch (2.x-release) is used for the built package. (See the
FAQs in the demo package for more explanation of that.)

symbioquine committed:
Update farm_map_google behavior contrib module for farmOS-map 2.x

**Why?** farmOS-map 2.x no longer contains core support for Google map
layers. Thus this contrib module will need to bundle that itself.

**What?** I'm converting this package to work similarly to
https://github.com/symbioquine/farmOS_vue_page_demo where
one branch (2.x) contains the unbuilt javascript and another
branch (2.x-release) is used for the built package. (See the
FAQs in the demo package for more explanation of that.)

paul121 commented on farm_map_google/farm_map_google.module:
```diff
@@ -17,7 +17,7 @@ function farm_map_google_library_info_build() {
   // Add a version.
   // See https://developers.google.com/maps/documentation/javascript/versions
-  $version_param = '?v=weekly';
+  $version_param = '?v=3';
```
Was this changed on purpose? I believe that version 3 was deleted and will default to weekly anyways. This is the current case for farmOS 1.x and we decided to just leave it: https://github.com/farmOS/farmOS/pull/383

There might be an argument for `quarterly` instead of `weekly`, curious your thoughts! I was also thinking this could be a configuration option for the module. I don't think it's a huge deal though.

symbioquine replied:
Good catch. I think I left this in by accident from when I was troubleshooting. Will change back.

paul121 commented on farm_map_google/farm_map_google.info.yml:
```diff
@@ -4,4 +4,4 @@ type: module
 package: farmOS
 core_version_requirement: ^9
 dependencies:
-  - farm:farm_map
```
Is there a reason for this change? I think dependencies are supposed to be prefixed by the drupal project name: https://www.drupal.org/node/2000204#s-complete-example

That said I do feel like I vaguely remember hearing a reason *not* to prefix by the project name, but I forget why.

symbioquine replied:
See https://www.drupal.org/project/farm/issues/3151247#comment-13956259

paul121 commented on .github/workflows/run-tests.yml:
```diff
@@ -0,0 +1,28 @@
+        run: docker-compose exec -u www-data -T www composer require paul121/farm_map_google
+      - name: Use drush to enable the farm_map_google module
+        run: docker-compose exec -u www-data -T www drush en farm_map_google
+      - name: Perform Minimal Validation That behavior JS loads
```
A bit of a nit pick, but it would be nice to test this with a user other than `root`. There have been a few bugs in farmOS core that we don't catch because we aren't testing with our "normal" farm user roles.

My preference would be to have something like this be a common practice: https://github.com/farmOS/farmOS/blob/dadb77e30f6b61645e010c3b2918db3bea831976/.tugboat/config.yml#L36

Would be nice to actually test this with phpunit but this is certainly better than nothing!!

symbioquine replied:
> A bit of a nit pick, but it would be nice to test this with a user other than `root`. There have been a few bugs in farmOS core that we don't catch because we aren't testing with our "normal" farm user roles.
> 
> My preference would be to have something like this be a common practice: https://github.com/farmOS/farmOS/blob/dadb77e30f6b61645e010c3b2918db3bea831976/.tugboat/config.yml#L36

On closer inspection, the user isn't relevant. Drupal doesn't require login to access JS resources so I'll just remove the OAUTH/authentication part.

> Would be nice to actually test this with phpunit but this is certainly better than nothing!!

Agreed at least in theory - though there isn't much to test here and I haven't changed the part that *could* be tested with phpunit.

Probably a better use of time would be writing some sort of end-to-end/integration test (perhaps with puppeteer) that validates the Google tiles are actually getting loaded, but I don't think I'm going to take on that or the adding the phpunit tests right now.

paul121 replied:
> On closer inspection, the user isn't relevant. Drupal doesn't require login to access JS resources so I'll just remove the OAUTH/authentication part.

Oops, yes I didn't look closely enough at the request path, apologies! That is correct

paul121 commented on .github/workflows/run-tests.yml:
```diff
@@ -0,0 +1,28 @@
+      - name: Wait until www container is ready
+        run: until [ -f ./www/www-container-fs-ready ]; do sleep 0.1; done && while { ! exec 3<>/dev/tcp/localhost/5432; } > /dev/null 2>&1; do sleep 0.1; done
+      - name: Do a site-install
+        run: docker-compose exec -u www-data -T www bash -c 'drush site-install farm --locale=en --db-url=pgsql://farm:farm@db/farm --site-name=Test0 --account-name=root --account-pass=test'
```
Might be good to install `all` modules here just in case something else in core were to cause a conflict. Like we do in the core tests: https://github.com/farmOS/farmOS/blob/dadb77e30f6b61645e010c3b2918db3bea831976/.github/workflows/run-tests.yml#L69

Also a nit pick, but could we change the admin credentials to `admin/admin`? Mainly because `admin` is the Drupal default username and seems to be a standard convention. Not a big deal tho..

symbioquine replied:
Both sound reasonable. Will change.

paul121 reviewed (commented):
Just some minor things I noticed. Overall looks good although I have not tested it out. But am glad that we will at least have some automated tests!

paul121 commented:
Because this starts to mention the package being on Packagist as `paul121/farm_map_google` maybe we could move this repository to `farmOS/farm_map_google` after merging  this in?

Or maybe we should just add it as a project on drupal.org and use the drupal.org composer repository? Start adding to our farmOS ecosystem https://www.drupal.org/project/farm/ecosystem :-)

Curious what you think @mstenta

mstenta commented:
I say we start by setting up drupal.org/project/farm_map_google!

As for farmOS/farm_map_google - that might be a bigger discussion/decision about whether or not farmOS (the "organization") will officially maintain it. I'm on the fence about that, personally, and might rather it be "community" maintained. The difference is subtle, of course, because it will still be you and me and @symbioquine maintaining it to start either way - but keeping it under the farmOS org adds future expectations - and by keeping it out of there we give ourselves a bit more flexibility.

symbioquine committed:
Address review feedback

symbioquine committed:
Couple more cleanup and doc improvement changes

symbioquine commented:
I think I've addressed all the feedback so far other than the ultimate org ownership/disposition. @paul121 Let me know if you see anything else.

@mstenta I'm not sure what the best way to proceed in the direction of putting this under the farmOS org, but I think that makes sense. One strategy I can image would be Paul merging this, then you forking the repo and me sending a follow-up PR to adjust the doc/test references for the fork.

symbioquine committed:
Update to new patched version of farmOS-olgm

symbioquine commented:
In speaking with @mstenta earlier today we roughly aligned on the course of action of not putting this contrib module under the Github `farmOS` org, and instead using a drupal.org project. The thought was that because @paul121 started this repo here, he should have the option to be the owner of the drupal.org project.

Since then, I've come up with a couple questions which I think we should resolve before proceeding with that strategy;

1. Personally, I feel the glue code in this PR (and the original repo here) should be MIT licensed like farmOS-map. If we are aligned on that, how do we determine if this code can then be hosted on drupal.org? I'm not sure I'm qualified to determine that from https://www.drupal.org/about/licensing
2. I've put some effort into figuring out this pattern of hosting the built/unbuilt JS in Github branches and using Github actions to publish the built code upon new tagged releases. I'd like confirmation from @mstenta / @paul121 that there isn't a better way to do it in the drupal.org world and - if so - that it's worth it for me to duplicate that effort to find a parallel pattern for automatically publishing built code to branches in drupal.org / gitlab just for the sake of avoiding moving this repo under the farmOS org on Github...

mstenta commented:
> Personally, I feel the glue code in this PR (and the original repo here) should be MIT licensed like farmOS-map

My understanding is that any Drupal module that's "dstributed" (shared with others), whether on Drupal.org or not, must be licensed as GPL, as it is legally considered a "derivative" work of Drupal.

> Drupal modules and themes are a derivative work of Drupal. If you distribute them, you must do so under the terms of GPL version 2 or later. 

https://www.drupal.org/docs/user_guide/en/understanding-gpl.html

Drupal.org enforces this by automatically adding a GPL `LICENSE.txt` file to any module that it packages.

> I'd like confirmation from @mstenta / @paul121 that there isn't a better way to do it in the drupal.org world and - if so - that it's worth it for me to duplicate that effort to find a parallel pattern for automatically publishing built code to branches in drupal.org

Hmm - this is a good question. And speaks to the larger problem of hosting farmOS modules on drupal.org (whether they be under the farmOS GitHub org or not)... AFAIK drupal.org doesn't offer the same abilities to run npm during packaging, so it may not be possible for drupal.org to host the "packaged" releases.

The primary benefit that hosting on drupal.org will provide is inclusion of translatable strings in localize.drupal.org - but we will probably need to host the packaged tarballs themselves on GitHub, much like we will be doing for farmOS itself (for other reasons... including: https://www.drupal.org/node/2715441).

However, this does mean that `composer require drupal/farm_map_google` wouldn't work, because it would not include the built JS files....

🤔

symbioquine commented:
> > Personally, I feel the glue code in this PR (and the original repo here) should be MIT licensed like farmOS-map
> 
> My understanding is that any Drupal module that's "dstributed" (shared with others), whether on Drupal.org or not, must be licensed as GPL, as it is legally considered a "derivative" work of Drupal.
> 
> > Drupal modules and themes are a derivative work of Drupal. If you distribute them, you must do so under the terms of GPL version 2 or later.
> 
> https://www.drupal.org/docs/user_guide/en/understanding-gpl.html
> 
> Drupal.org enforces this by automatically adding a GPL `LICENSE.txt` file to any module that it packages.

I think in re-reading https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses more carefully, the MIT license that I'm checking in with this PR should be one of those GPL compatible ones since the text is identical to the so called [X11 Lincense](https://directory.fsf.org/wiki/License:X11) minus the X trademark verbiage.

Further, from https://www.drupal.org/about/licensing#gpl-compatible

> ![image](https://user-images.githubusercontent.com/30754460/122782762-d6128580-d265-11eb-9f3b-8132790906ce.png)

It looks like there shouldn't be a problem licensing the JS parts of this PR as MIT. I suspect that if the built code - including both the GPL module code and the MIT JS code - were then distributed through drupal.org the JS parts would end up dual licensed as both GPL and MIT depending where one acquires the JS code.

If this premise is correct, I believe all I would need to do is add a GPL LICENSE.txt file to the `farm_map_google/` directory of this repo and a note to the README disambiguating that the Drupal module portion of the code is GPL licensed and the rest is MIT licensed.

> > I'd like confirmation from @mstenta / @paul121 that there isn't a better way to do it in the drupal.org world and - if so - that it's worth it for me to duplicate that effort to find a parallel pattern for automatically publishing built code to branches in drupal.org
> 
> Hmm - this is a good question. And speaks to the larger problem of hosting farmOS modules on drupal.org (whether they be under the farmOS GitHub org or not)... AFAIK drupal.org doesn't offer the same abilities to run npm during packaging, so it may not be possible for drupal.org to host the "packaged" releases.
> 
> The primary benefit that hosting on drupal.org will provide is inclusion of translatable strings in localize.drupal.org - but we will probably need to host the packaged tarballs themselves on GitHub, much like we will be doing for farmOS itself (for other reasons... including: https://www.drupal.org/node/2715441).
> 
> However, this does mean that `composer require drupal/farm_map_google` wouldn't work, because it would not include the built JS files....
> 
> thinking

Another option we may be able to consider is keeping the original source on Github (the ownership/org isn't that important) and using a Github action to publish the built code to drupal.org. I haven't tested it yet, but it looks like it is possible to configure access tokens or deploy keys in the drupal.org Gitlab. I believe either of those could be added as secrets in Github and used to push the built code to drupal.org from a Github action.

@mstenta What you think about my understanding/premise on the licensing side of things?

Have heard of anyone pushing code/artifacts automatically to drupal.org from other places?

mstenta commented:
> @mstenta What you think about my understanding/premise on the licensing side of things?

That makes sense to me! (With the obligatory "I am not a lawyer") :-)

> Have heard of anyone pushing code/artifacts automatically to drupal.org from other places?

I haven't - but I'd be curious what's possible!

One thing I do know is that drupal.org maintains a "Library Packaging Whitelist": https://www.drupal.org/project/drupalorg_whitelist

This is specifically used when packaging installation profiles / distributions, which need to pull in third-party libraries via Drush Make (and so this really only applied to farmOS 1.x - but Composer-based packaging is not supported yet).

The whitelist provided a mechanism for drupal.org to ensure that only GPL-compatible licensed code could be included in packaged distros hosted on drupal.org.

And on that note... yes: MIT is GPL compatible. I helped to get a few libraries added to the whitelist in the past (including farmOS-map itself):

- https://www.drupal.org/project/drupalorg_whitelist/issues/3114971
- https://www.drupal.org/project/drupalorg_whitelist/issues/3126545
- https://www.drupal.org/project/drupalorg_whitelist/issues/2933381

paul121 commented:
> The thought was that because @paul121 started this repo here, he should have the option to be the owner of the drupal.org project.

I appreciate the thought but must say I don't have much attachment to this repo! I only started it as the first step in porting the Google maps features from 1.x core to contrib. Especially with the Google maps behavior being removed from farmOS-map and to-be included here, I feel that this is creeping a bit out of my wheelhouse (which is fine!)

I should also note that I'm not using this module myself or for any projects. So I really have little incentive to be a good maintainer :P

symbioquine commented:
> > Have heard of anyone pushing code/artifacts automatically to drupal.org from other places?
> 
> I haven't - but I'd be curious what's possible!

Cool. I'll see whether I can test out the use of repo access keys/tokens to automatically push to my farmos_wfs module... It would be cool if that were installable with `composer require drupal/farmos_wfs` :)

> One thing I do know is that drupal.org maintains a "Library Packaging Whitelist": https://www.drupal.org/project/drupalorg_whitelist
> 
> This is specifically used when packaging installation profiles / distributions, which need to pull in third-party libraries via Drush Make (and so this really only applied to farmOS 1.x - but Composer-based packaging is not supported yet).
> 
> The whitelist provided a mechanism for drupal.org to ensure that only GPL-compatible licensed code could be included in packaged distros hosted on drupal.org.

Good to know, though I suspect it wouldn't be relevant in the case of pushing built code to a drupal.org Gitlab repo.

symbioquine commented:
> > > Have heard of anyone pushing code/artifacts automatically to drupal.org from other places?
> > 
> > 
> > I haven't - but I'd be curious what's possible!
> 
> Cool. I'll see whether I can test out the use of repo access keys/tokens to automatically push to my farmos_wfs module... It would be cool if that were installable with `composer require drupal/farmos_wfs` :)

Well, it took a while to get it working, but I've managed to make my [farmOS_wfs](https://github.com/symbioquine/farmOS_wfs) module automatically push new commits/tags on the 2.x branch to the [Drupal.org Gitlab repo](https://git.drupalcode.org/project/farmos_wfs).

The code for that is here: https://github.com/symbioquine/farmOS_wfs/blob/2.x/.github/workflows/push-to-drupal-dot-org.yml

Some learnings:

* There doesn't seem to be an easy way to push all tags and branches over due to the kinds of refs that the Gitlab instance rejects
* Version tags for releases can't have a 'v' prefix - e.g. 'v0.1.4'
* Version tags that are <1.0.0 don't show up right when you create releases
